ui/components/institutional/interactive-replacement-token-notification/interactive-replacement-token-notification.js

```diff
@@ -0,0 +1,139 @@
+import { getIsUnlocked } from '../../../ducks/metamask/metamask';
+import { useI18nContext } from '../../../hooks/useI18nContext';
+import { mmiActionsFactory } from '../../../store/institutional/institution-background';
+import { sha256 } from '../../../../shared/modules/hash.utils';
+import {
+  Size,
+  IconColor,
+  AlignItems,
+  DISPLAY,
+  BLOCK_SIZES,
+  JustifyContent,
+  TextColor,
+  BackgroundColor,
+} from '../../../helpers/constants/design-system';
+import {
+  Icon,
+  IconName,
+  IconSize,
+  ButtonLink,
+  Text,
+} from '../../component-library';
+
+import Box from '../../ui/box';
+
+const InteractiveReplacementTokenNotification = ({ isVisible }) => {
+  const t = useI18nContext();
+  const dispatch = useDispatch();
+  const mmiActions = mmiActionsFactory();
+
+  const keyring = useSelector(getCurrentKeyring);
+  const address = useSelector(getSelectedAddress);
+  const isUnlocked = useSelector(getIsUnlocked);
+  const interactiveReplacementToken = useSelector(
+    getInteractiveReplacementToken,
+  );
+
+  const [showNotification, setShowNotification] = useState(isVisible);
+
+  const handleShowNotification = async () => {
+    const hasInteractiveReplacementToken =
+      interactiveReplacementToken &&
+      Boolean(Object.keys(interactiveReplacementToken).length);
+
+    if (!/^Custody/u.test(keyring.type)) {
+      console.log('You are not a custodian');
+      setShowNotification(false);
+      return;
+    } else if (!hasInteractiveReplacementToken) {
+      console.log("You don't have an interactive replacement token");
+      setShowNotification(false);
+      return;
+    }
+
+    const token = await dispatch(mmiActions.getCustodianToken());
+    const custodyAccountDetails = await dispatch(
+      mmiActions.getAllCustodianAccountsWithToken(
+        keyring.type.split(' - ')[1],
+        token,
+      ),
+    );
+
+    const showNotificationValue =
+      isUnlocked &&
+      interactiveReplacementToken.oldRefreshToken &&
+      custodyAccountDetails &&
+      Boolean(Object.keys(custodyAccountDetails).length);
+
+    let tokenAccount;
+
+    if (Array.isArray(custodyAccountDetails)) {
+      tokenAccount = custodyAccountDetails
+        .filter((item) => item.address.toLowerCase() === address.toLowerCase())
+        .map((item) => ({
+          token: item.authDetails?.refreshToken,
+        }))[0];
+    }
+
+    const refreshTokenAccount = await sha256(
+      tokenAccount?.token + interactiveReplacementToken.url,
+    );
+
+    console.log('showNotificationValue', showNotificationValue);
+    console.log(
+      'has the same old token?',
+      refreshTokenAccount === interactiveReplacementToken.oldRefreshToken,
+    );
+
+    setShowNotification(
+      showNotificationValue &&
+        refreshTokenAccount === interactiveReplacementToken.oldRefreshToken,
+    );
+  };
+
+  useEffect(() => {
+    handleShowNotification();
```

Comment:
handleShowNotification references a state key. handleShowNotification is recreated on every render, but the effect is only created once. This means that the version of handleShowNotification that is created on the first render will never be garbage collected and any values that end up getting tied up in the function body of handleShowNotification will also not be released and possible future fibers created might not be garbage collected if they point to values in the first fiber. 

All of this to say you should declare this function *inside* the useEffect. This will force you to declare dependencies inside the useEffect.

Reply:
Indeed 👍🏼 , thank you for explaining it, will change that.

Reply:
This is likely throwing a warning in lint now, please resolve.

Reply:
https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/42692/workflows/09f3c544-a1fe-49b3-bb31-cf2f348530e7/jobs/1196437?invite=true#step-103-7